From e948b702e1c311e5efb2e3a7b20cef53ee14cf0e Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 3 Apr 2026 18:01:12 +0800
Subject: [PATCH 36/98] backend-drm: Fix error when reinitializing renderer

Reset the plane state before destroying renderer buffers.

This reverts part of commit 853aa1c5 ("backend-drm: destroy_sprites()
after outputs").

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/drm-gbm.c | 14 +++++++++++---
 libweston/backend-drm/drm.c     |  8 ++++++--
 2 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/libweston/backend-drm/drm-gbm.c b/libweston/backend-drm/drm-gbm.c
index f3caa035..711d2c12 100644
--- a/libweston/backend-drm/drm-gbm.c
+++ b/libweston/backend-drm/drm-gbm.c
@@ -755,8 +755,12 @@ drm_output_fini_egl(struct drm_output *output)
 	const struct weston_renderer *renderer = b->compositor->renderer;
 
 	/* Destroying the GBM surface will destroy all our GBM buffers,
-	 * regardless of refcount. */
-	weston_assert_ptr_null(b->compositor, output->scanout_handle);
+	 * regardless of refcount. Ensure we destroy them here. */
+	if (!b->compositor->shutting_down && output->scanout_handle &&
+	    output->scanout_handle->plane->state_cur->fb &&
+	    output->scanout_handle->plane->state_cur->fb->type == BUFFER_GBM_SURFACE) {
+		drm_plane_reset_state(output->scanout_handle->plane);
+	}
 
 	renderer->gl->output_destroy(&output->base);
 	gbm_surface_destroy(output->gbm_surface);
@@ -770,7 +774,11 @@ drm_output_fini_vulkan(struct drm_output *output)
 	struct drm_backend *b = output->backend;
 	const struct weston_renderer *renderer = b->compositor->renderer;
 
-	weston_assert_ptr_null(b->compositor, output->scanout_handle);
+	if (!b->compositor->shutting_down && output->scanout_handle &&
+	    output->scanout_handle->plane->state_cur->fb &&
+	    output->scanout_handle->plane->state_cur->fb->type == BUFFER_DMABUF_BACKEND) {
+		drm_plane_reset_state(output->scanout_handle->plane);
+	}
 
 	for (unsigned int i = 0; i < ARRAY_LENGTH(output->renderbuffer); i++)
 		renderer->destroy_renderbuffer(output->renderbuffer[i]);
diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index db4aee0e..f99cc7cc 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -2037,8 +2037,12 @@ drm_output_fini_pixman(struct drm_output *output)
 	unsigned int i;
 
 	/* Destroying the Pixman surface will destroy all our buffers,
-	 * regardless of refcount. */
-	weston_assert_ptr_null(b->compositor, output->scanout_handle);
+	 * regardless of refcount. Ensure we destroy them here. */
+	if (!b->compositor->shutting_down && output->scanout_handle &&
+	    output->scanout_handle->plane->state_cur->fb &&
+	    output->scanout_handle->plane->state_cur->fb->type == BUFFER_PIXMAN_DUMB) {
+		drm_plane_reset_state(output->scanout_handle->plane);
+	}
 
 	for (i = 0; i < ARRAY_LENGTH(output->dumb); i++) {
 		renderer->destroy_renderbuffer(output->renderbuffer[i]);
-- 
2.39.5

